### PR TITLE
[pipeline] add pipeline support for T5Stack/T5EncoderModel

### PR DESCRIPTION
**Pull request by Fridge003**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

#4247

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

- Design special algorithm for distributing layers to pipeline stages, since the structure of T5(Encoder/Decoder) is more complex than other transformer models.
- Modify the logic of forward for T5Stack/T5EncoderModel, so that pipeline parallel can be smoothly run.
- Add relevant tests for layer distribution & pipeline forward.

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/shardformer/policies/t5.py [hunk 1]
```diff
@@ -204,19 +337,55 @@ def module_policy(self):
                                                         target_key=T5ForConditionalGeneration)
         return policy
 
+    def postprocess(self):
+        super().postprocess()
+        if self.shard_config.enable_tensor_parallelism:
```
Comment (CjhHa1) [inline colossalai/shardformer/policies/t5.py:342]: you should check pipeline parallelism

File: tests/test_shardformer/test_model/test_t5_pipeline_utils.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
+from colossalai.shardformer.policies.t5 import T5BasePolicy
+
```
Comment (CjhHa1) [inline tests/test_shardformer/test_model/test_t5_pipeline_utils.py:2]: I prefer move this to test_pipeline as all policy related tests are in.

**Review by CjhHa1 — APPROVED** [review]
LGTM

**Comment by Fridge003** [comment]
截屏2023-07-21 14 59 58

**Pull request merged**